tests/test_simplify.py

```diff
@@ -1,10 +1,13 @@
+from logging import root as root_logger, DEBUG
 from pathlib import Path
 
 import pytest
 import pyfqmr
 
 import numpy as np
 
+root_logger.setLevel(DEBUG)
```

Comment:
I'm really not comfortable with this. If someone calls `simplify_mesh(..., verbose=True)` then, unless they know to set the logging level to debug (which they can only find from reading the source code), then nothing will happen.

Are you really not OK with leaving `verbose=False` and doing your own informational output?

Reply:
I'm happy with using `logger.info` instead of `logger.debug`, just in my mind "verbose" means "debug".

Using logging is a lot more controllable, and allows the user to disable if they don't want.

Also the simplify library already logs out the number of faces.

Finally, the details you log out are pretty easily figured out by the user

Reply:
Ahh, I missed that `verbose=True` was the default. In that case I feel like all options are wrong options.

Reply:
Would it be too confusing to set `verbose` to the logging level? Then make the default value of `verbose` be `logging.INFO`?

Reply:
You mean to do this?

```cython
from logging import getLogger, INFO

cdef simplify_mesh(..., verbose: int = INFO, ...):
    ...

    simplify_mesh(..., verbose, ...)

    if verbose:
        logger = getLogger("pyfqmr")
        logger.log(verbose, 'simplified mesh in ...')
```

The problems with that are:

* That's a breaking API change, where existing users passing `True` or `False` will cause `logger.log` to raise an exception
* In the call to the Simplify's `simplify_mesh`, `verbose` must be a boolean, so I would suppose you do `verbose <= INFO`
* Generally, the message defines the log level, not the users

Honestly, I think the proper solution is to replace the `verbose` parameter to Simplify's `simplify_mesh` with a parameter which maybe-null pointer to a function to call with the log message, and then in pyfqmr the message is logged using the logger.

Reply:
> Honestly, I think the proper solution is to replace the verbose parameter to Simplify's simplify_mesh with a parameter which maybe-null pointer to a function to call with the log message, and then in pyfqmr the message is logged using the logger.

Would that even be possible for the `printf()`s in the cpp file?

Reply:
Would just flipping `verbose`'s default be such a bad thing? I can't imagine anyone being too upset over some informational print statements disappearing in a `0.x` version.

Reply:
> Would that even be possible for the printf()s in the cpp file?

No, it would require a minor code change: replace `printf` with 

```c
if (log != NULL) {
    char[50] message;
    snprintf(message, 50, ...);
    log(message, 50);
}
```

(Orr whatever instead of 50).

---

> Would just flipping verbose's default be such a bad thing?

I think that would be fine

Reply:
Can we delay the `import logging` again? I'd rather have no verbose mode at all than lengthen startup time.

Reply:
> I will do that, but if you're worried about startup time, you should remove the NumPy import. The logging import is pretty insignificant anyway.

Well, numpy has to be loaded at some point whereas it's possible for someone to use this library without logging anything at all. (And yes, I probably am just preach purity rather than practicality at this point. Guess I'm overcompensating for all the `from vtk import *`s I've seen.)

> Also, I thought you wanted to switch the default for verbose.

To be honest, I'm not really sure what I want. If I was building this:
* `verbose` would default to `False` where all logging would be skipped.
* `verbose=some_callable` would pass the informational strings to `some_callable()` (which you could set to `logging.debug()`). That removes any assumptions about preferred console outputs, logging levels or even that this is a console enabled application (this could be a GUI).
* `verbose=True` would be interpreted as `verbose=print`.

Reply:
Regarding the implementation of these changes, I don't have much to add as I haven't actively used this project in years. For API changes, I'd suggest following your own inputs, @bwoodsend.

As for the printing, it doesn't seem to provide much value and could either be removed or moved entirely into the C++ code. However, unless there are significant errors or segfaults that require detailed logging, that might be overkill. The mesh validity could still be checked after the simplification.

Additionally, exposing more C++ functions (such as [simplify_mesh_lossless](https://github.com/Kramer84/pyfqmr-Fast-Quadric-Mesh-Reduction/blob/d70abbfe3458d06b88040a0736d52c6fb3c49e18/pyfqmr/Simplify.h#L455)) could be an interesting addition, but perhaps that's for a future issue. On a related note, this code has also been adapted recently in [pyvista/fast-simplification](https://github.com/pyvista/fast-simplification), and there may be ideas worth borrowing from there.

By the way, @bwoodsend, since I don't collaborate often on GitHub, I just wanted to confirm—do you need my input or validation for pull requests like these? If not, feel free to proceed as you see fit!

Reply:
`logging` now imported lazily. Note that the import system caches the import, so subsequent imports are barely anything more than a `dict` lookup

Reply:
> I haven't actively used this project in years.

Me neither. I originally wanted it because I was desperate to get away from using VTK for visualisation but knew I'd not be able to if I still required `vtkQuadricDecimation()`. But I never did find a VTK alternative for the visualisations though so I never got round to actually using this library. Otherwise, I'd have likely done more to it (e.g. move from Cython to ctypes, add proper handling of bogus inputs so that this library on a server wouldn't be a security bomb, ...).

> the same underlying c++ code has also been adapted recently in [pyvista/fast-simplification](https://github.com/pyvista/fast-simplification), and there may be ideas worth borrowing from there.

I have no intention of entering a feature race. Unless that library does something dumb like forcing you to use pyvista then I'd be tempted to throw in the towel and call it our successor.

Reply:
Thanks Laurie, that'll do